File: codebase2/backend/app/ai/services/concept_extractor.py
"""Deterministic teaching-context extraction."""

import re

STOP_WORDS = {
    "about", "after", "also", "and", "are", "based", "before", "being", "between", "can", "concept", "context", "does", "for", "from", "into", "lesson", "more", "must", "not", "that", "the", "this", "through", "using", "with", "what", "which", "your",
    "các", "cho", "của", "được", "khái", "là", "một", "này", "những", "trong", "và", "về",
}


def _clean(value: object) -> str:
    return re.sub(r"\s+", " ", str(value or "")).strip()


def _title_case(value: str) -> str:
    return value[:1].upper() + value[1:]


def extract_concepts(source_text: object = "", title: object = "", source_id: object = "") -> dict[str, object]:
    """Derive a small deterministic teaching context that a later service may replace."""
    text = _clean(source_text)
    heading = _clean(title)
    if not text and not heading:
        raise ValueError("Concept extraction requires sourceText or title.")
    candidates = [heading] if heading else []
    frequency: dict[str, int] = {}
    for word in re.findall(r"[^\W_][\w-]{2,}", f"{heading} {text}".lower(), flags=re.UNICODE):
        if word not in STOP_WORDS:
            frequency[word] = frequency.get(word, 0) + 1
    for word, _ in sorted(frequency.items(), key=lambda pair: (-pair[1], pair[0])):
        if not any(item.lower() == word for item in candidates):
            candidates.append(_title_case(word))
        if len(candidates) >= 3:
            break
    concepts = candidates[:3]
    primary_concept = concepts[0] if concepts else "Teaching concept"
    return {
        "topic": heading or primary_concept,
        "concepts": concepts,
        "learningObjectives": [f"Explain {primary_concept} using the supplied teaching material."],
        "sourceId": _clean(source_id) or None,
    }

File: codebase/ai/services/conceptExtractor.js
'use strict';

const STOP_WORDS = new Set(['about', 'after', 'also', 'and', 'are', 'based', 'before', 'being', 'between', 'can', 'concept', 'context', 'does', 'for', 'from', 'into', 'lesson', 'more', 'must', 'not', 'that', 'the', 'this', 'through', 'using', 'with', 'what', 'which', 'your', 'các', 'cho', 'của', 'được', 'khái', 'là', 'một', 'này', 'những', 'trong', 'và', 'về']);

function clean(value) { return String(value || '').replace(/\s+/g, ' ').trim(); }
function titleCase(value) { return value.replace(/\b\p{L}/gu, character => character.toLocaleUpperCase()); }

/** Deterministically derive a small teaching context; it can later be replaced by an LLM implementation. */
function extractConcepts({ sourceText = '', title = '', sourceId = '' } = {}) {
  const text = clean(sourceText);
  const heading = clean(title);
  if (!text && !heading) throw new Error('Concept extraction requires sourceText or title.');
  const candidates = [];
  if (heading) candidates.push(heading);
  const frequency = new Map();
  for (const word of `${heading} ${text}`.toLocaleLowerCase().match(/[\p{L}][\p{L}\p{N}-]{2,}/gu) || []) {
    if (!STOP_WORDS.has(word)) frequency.set(word, (frequency.get(word) || 0) + 1);
  }
  for (const [word] of [...frequency.entries()].sort((a, b) => b[1] - a[1] || a[0].localeCompare(b[0]))) {
    if (!candidates.some(item => item.toLocaleLowerCase() === word)) candidates.push(titleCase(word));
    if (candidates.length >= 3) break;
  }
  const concepts = candidates.slice(0, 3);
  const primaryConcept = concepts[0] || 'Teaching concept';
  return {
    topic: heading || primaryConcept,
    concepts,
    learningObjectives: [`Explain ${primaryConcept} using the supplied teaching material.`],
    sourceId: clean(sourceId) || null
  };
}

module.exports = { extractConcepts };
